### PR TITLE
added support for inverse of 3x3 matrix

### PR DESCRIPTION
**Pull request by kituuu**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: matrix/inverse_of_matrix.py [hunk 1]
```diff
@@ -25,24 +26,137 @@ def inverse_of_matrix(matrix: list[list[float]]) -> list[list[float]]:
...
+
+    >>> inverse_of_matrix([[1, 2, 1], [0,3, 4]])
+    Traceback (most recent call last):
+    ...
+    ValueError: Please provide a matrix of size 2x2 or 3x3.
+
+    >>> inverse_of_matrix([[1, 2, 3], [7, 8, 9], [7, 8, 9]])
+    Traceback (most recent call last):
+    ...
+    ValueError: This matrix has no inverse.
```
Comment (cclauss) [inline matrix/inverse_of_matrix.py:63]: We should test that all lines between `Traceback` and the Error are indented at least four spaces.
```suggestion
        ...
    ValueError: This matrix has no inverse.

    >>> inverse_of_matrix([])
    Traceback (most recent call last):
        ...
    ValueError: Please provide a matrix of size 2x2 or 3x3.

    >>> inverse_of_matrix([[],[]])
    Traceback (most recent call last):
        ...
    ValueError: Please provide a matrix of size 2x2 or 3x3.

    >>> inverse_of_matrix([[1, 2], [3, 4], [5, 6]])
    Traceback (most recent call last):
        ...
    ValueError: Please provide a matrix of size 2x2 or 3x3.

    >>> inverse_of_matrix([[1, 2, 1], [0,3, 4]])
    Traceback (most recent call last):
        ...
    ValueError: Please provide a matrix of size 2x2 or 3x3.

    >>> inverse_of_matrix([[1, 2, 3], [7, 8, 9], [7, 8, 9]])
    Traceback (most recent call last):
        ...
    ValueError: This matrix has no inverse.
```

File: matrix/inverse_of_matrix.py [hunk 2]
```diff
@@ -25,24 +26,137 @@ def inverse_of_matrix(matrix: list[list[float]]) -> list[list[float]]:
...
-        raise ValueError("This matrix has no inverse.")
+        # Transpose the cofactor matrix (Adjoint matrix)
+        adjoint_matrix = [
+            [d(0.0), d(0.0), d(0.0)],
+            [d(0.0), d(0.0), d(0.0)],
+            [d(0.0), d(0.0), d(0.0)],
+        ]
+        for i in range(3):
+            for j in range(3):
+                adjoint_matrix[i][j] = cofactor_matrix[j][i]
```
Comment (cclauss) [inline matrix/inverse_of_matrix.py:151]:
```suggestion
        adjoint_matrix = zip(*cofactor_matrix)
```
https://www.geeksforgeeks.org/transpose-matrix-single-line-python

File: matrix/inverse_of_matrix.py [hunk 3]
```diff
@@ -25,24 +26,137 @@ def inverse_of_matrix(matrix: list[list[float]]) -> list[list[float]]:
...
+            [float(d(n) / determinant) or 0.0 for n in row] for row in swapped_matrix
+        ]
+    elif (
+        len(matrix) == 3
+        and len(matrix[0]) == 3
+        and len(matrix[1]) == 3
+        and len(matrix[2]) == 3
+    ):
+        # Calculate the determinant of the matrix using Sarrus rule
+        determinant = (
```
Comment (cclauss) [inline matrix/inverse_of_matrix.py:97]:
```suggestion
        determinant = float(

```

File: matrix/inverse_of_matrix.py [hunk 4]
```diff
@@ -25,24 +26,137 @@ def inverse_of_matrix(matrix: list[list[float]]) -> list[list[float]]:
...
 
-    # Creates a copy of the matrix with swapped positions of the elements
-    swapped_matrix = [[0.0, 0.0], [0.0, 0.0]]
-    swapped_matrix[0][0], swapped_matrix[1][1] = matrix[1][1], matrix[0][0]
-    swapped_matrix[1][0], swapped_matrix[0][1] = -matrix[1][0], -matrix[0][1]
+        # Inverse of the matrix using the formula (1/determinant) * adjoint matrix
+        inverse_matrix = [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]]
+        for i in range(3):
+            for j in range(3):
+                inverse_matrix[i][j] = float(adjoint_matrix[i][j]) / float(determinant)
```
Comment (cclauss) [inline matrix/inverse_of_matrix.py:157]:
```suggestion
                inverse_matrix[i][j] /= determinant
```

File: matrix/inverse_of_matrix.py (hunk 2 above)
Comment (kituuu) [reply matrix/inverse_of_matrix.py:151]: Ok sir.

File: matrix/inverse_of_matrix.py (hunk 3 above)
Comment (kituuu) [reply matrix/inverse_of_matrix.py:97]: I am on it sir.

**Comment by kituuu** [comment]
@cclauss Sir, the zip() method involves a list of tuples. And in my code, every matrix was a list of lists. So I used NumPy.array instead of zip. Also, I don't know what error doctest is giving, can u help me with that?

**Comment by kituuu** [comment]
@poyea @cclauss @dhruvmanila Sir, please review this also. And suggest changes if required. 🥲

File: matrix/inverse_of_matrix.py [hunk 5]
```diff
@@ -2,47 +2,162 @@
...
+        for i in range(3):
+            for j in range(3):
+                inverse_matrix[i][j] /= d(determinant)
 
-    # Calculate the inverse of the matrix
-    return [[float(d(n) / determinant) or 0.0 for n in row] for row in swapped_matrix]
+        # Calculate the inverse of the matrix
+        return [[float(d(n)) or 0.0 for n in row] for row in inverse_matrix]
+    else:
+        raise ValueError("Please provide a matrix of size 2x2 or 3x3.")
```
Comment (ChrisO345) [inline matrix/inverse_of_matrix.py:163]:
```suggestion
    raise ValueError("Please provide a matrix of size 2x2 or 3x3.")
```

Comment (cclauss) [reply matrix/inverse_of_matrix.py:163]: For me(old school) this change is optional.

**Review by cclauss — APPROVED** [review]
I applied all code review suggestions.

Thanks for your persistence!!

**Pull request merged**